operator/docs/enhancements/ruler_support.md

```diff
@@ -0,0 +1,686 @@
+
+#### General constraints
+
+1. The above `LokiRulerConfig` is limited to a single global remote write endpoint for exporting metrics from recording rules. This leaves solving multi-tenancy issues on the remote write server side and provide means (e.g. headers) for spliting/amending metrics ingestion per tenant. On the other hand it simplifies ruler operations as it does not require to spin concurrent remote write clients per tenant.
+2. Additionally the `LokiRulerConfig** requires a user-provided Kubernetes Secret for sensitive information. This adds an extra dependency that requires validation inside the controller-loop.
+
+### Risks and Mitigations
+
+#### Loki Ruler Configuration versioning
+
+**Risk**: The `LokiRulerConfig` CRD exposes an almost identical [ruler config](https://grafana.com/docs/loki/latest/configuration/#ruler) to support AlertManager and Remote-Write. Both sub-specifications are subject of change to the API version of the server endpoint (i.e. AlertManager, Remote-Write). Although AlertManager is using a versioned approach e.g. `EnableV2` switch, we miss a similar approach accross remote write endpoints.
+
+**Mitigation**: We require a support matrix for `LokiRulerConfig` versions to AlertManager API versions and Remote Write implementations. For later we could use a list of releases (e.g. Prometheus 2.x, Thanos 0.20.z).
+
+#### Single CRD for Alerting and Recording Rules
+
+**Risk**: `LokiRule` is an amalgam of types for alerting and recording rules. This requires custom validation to inform the user of bad inputs and in turn extra maintaince over time. In addition alerting rules might deviate with new features in future (See [Feature Request: alert relabel configs](https://github.com/grafana/loki/issues/5886)). Thus a CRD combining both rules types might become confusing and error-prone from a user experience perspective.
+
+**Mitigation**: Either split the two types in an `AlertingRule` and `RecordingRule` type now or in a future v2 version. Later would probably be harder to support seamless migrations from the Kubernetes API server side.
+
+#### Single ConfigMap for all LokiRule instances
+
+**Risk**: The proposed reconciliation approach manifests that all `LokiRule` instances are transformed to individual entries in a single ConfigMap. Alerting/Recording rules can be become quite big (e.g. large LogQL expresssions). As per ConfigMaps represent Kubernetes resource stored in etcd, this might hit some store limits (See [1MiB limit](https://kubernetes.io/docs/concepts/configuration/configmap/#motivation)). Therefore storing all rules in a single ConfigMap might become impossible over time or on large clusters.
+
+**Mitigation**: Introduce some sort of sharding of LokiRules into multiple ConfigMap resources.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Do we need to split the `LokiRule` into two types `RecordingRule` and `AlertingRule`?
```

Comment:
This sounds like a good idea to me. Currently there is the possibility to specify both `Alert` and `Record` in a rule.

Reply:
I agree, if you can, provide two distinct types.
I don't know the history or any arguments that made Prometheus go with one "type" (in quotes since for prometheus its just yaml) that can be rule or alert, so I might be missing upsides of this.
 
One downside we are experiencing is that when one would like to give users the option to create alerting rules, but explicitly not recording rules. With a type that can be either one has to resort to either a validating webhook which muddies the user experience a bit or one ends up creating a new type. See https://github.com/openshift/enhancements/pull/1030#discussion_r854971161 for more context.

The ability to specify which rule variant one expects can be very handy. And you'll probably end up with a union type anyway.

Reply:
Thanks for the feedback @jan--f . I have split the `LokiRule` type into an `AlertingRule` and a `RecordingRule` types. This brings a lot of clarity in using the APIs in distinct scenarios. Plus it helps users to apply different RBAC on each one.

Reply:
Eh, forgot to submit my feedback... too late 3c:

I think in some way it makes sense to keep the semantics of https://github.com/prometheus-operator/prometheus-operator PrometheusRule. At the end simplicity of the API is important and less components = the better. Also the ecosystem got use to this, so it's might be easier to use.

@periklis @jan--f  the history is most likely coming from Prometheus itself: one type is easier than two and both types are somewhat "rules" (one is recording, second alering). In Prometheus we use same rules manager, same rules API (/api/v1/rules) for both on Prometheus given similar semantics (`expr`).

I think both ways work, but consistency is key here. It's weird I can list all my `lokiAlertRule` and `lokiRecordRule` in Kube API for Loki, but I have only `PrometeusRule` in Prom operator. It would be amazing to move either way for both projects (: (and ideally have Prom Alert easily convertible to Loki Alert for ease of use and so on...)

Reply:
> Eh, forgot to submit my feedback... too late 3c:
> 
> I think in some way it makes sense to keep the semantics of https://github.com/prometheus-operator/prometheus-operator PrometheusRule. At the end simplicity of the API is important and less components = the better. Also the ecosystem got use to this, so it's might be easier to use.
> 
> @periklis @jan--f the history is most likely coming from Prometheus itself: one type is easier than two and both types are somewhat "rules" (one is recording, second alering). In Prometheus we use same rules manager, same rules API (/api/v1/rules) for both on Prometheus given similar semantics (`expr`).

This will continue to be the case with Loki if you use the Rules API. You will be able to retrieve both types of rules from the Loki. The `AlertingRule` and `RecordingRule` CRD are just a storage form for k8s and both get translated into rules that Loki get read, e.g. examplary output from cortex tool:

```
Namespace                                                                 | Rule Group
openshift-logging-lokistack-dev-17f964d8-35bc-4586-89d2-4f071b36c090.yaml | should_record
openshift-logging-lokistack-dev-623fa43e-3083-450c-959d-db575016bc33.yaml | should_fire
openshift-logging-lokistack-dev-623fa43e-3083-450c-959d-db575016bc33.yaml | credentials_leak
```
where `should_record` is a k8s `RecordingRule` and `should_fire` is a k8s `AlertingRule`

> 
> I think both ways work, but consistency is key here. It's weird I can list all my `lokiAlertRule` and `lokiRecordRule` in Kube API for Loki, but I have only `PrometeusRule` in Prom operator. It would be amazing to move either way for both projects (: (and ideally have Prom Alert easily convertible to Loki Alert for ease of use and so on...)

I would say this is a call for the prometheus-operator to re-consider splitting into two types in future. I won't say consistency is bad or neglible, but AFAICS and as @jan--f mentions having RBAC-control on different types is something that users want to have. For example alerts are simple and can on misuse just be noisy on firing. OTH recording rules need attention before processing because they might overload a remote write endpoint. Thus admins would possible want to limit who can create recording rules, while leave alerting rules for general use free.